### PR TITLE
Final flake8 PR...

### LINKED ISSUE #13525: Use __all__ top-level variable in module files
I have seen `__all__` top-level variable set to list of objects that are imported when we do `from sympy.combinatorics.coset_table import *` I get `bisect_left` which isn't even supposed to be in the module.

Also I think this will reduce import time when doing `from sympy.combinatorics.coset_table import *` (will this also reduce import time for `from sympy import *`? after considering the files like `coset_table.py` has public available classes and functions)

I have seen this often used in scikit-learn source code. This will be easy to fix, once someone comments that it is a good decision to have this.

[Edit]: I know the existence of `__all__` in some of module `__init__` files, but still a large number of them are missing, ex. combinatorics module.

### LINKED ISSUE #13592: "from sympy import *" imports every package name
I noticed that `from sympy import *` additionally imports every package name, like `core` or `functions`. Is it an intended behavior?

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Fixes https://github.com/sympy/sympy/issues/13525
Fixes https://github.com/sympy/sympy/issues/13592

#### Release Notes

- other
  - Submodule names are no longer imported with `from sympy import *`. They can still be imported directly like `from sympy import core` or accessed like `sympy.core`, or like `sys.modules['sympy.simplify']` for modules that share names with SymPy functions.

### CODE REVIEW

File: sympy/__init__.py [hunk 1]
```diff
@@ -54,40 +54,424 @@ def __sympy_debug():
...
+
+    'euler_equations', 'singularities', 'is_increasing',
+    'is_strictly_increasing', 'is_decreasing', 'is_strictly_decreasing',
+    'is_monotonic', 'finite_diff_weights', 'apply_finite_diff',
+    'as_finite_diff', 'differentiate_finite', 'periodicity', 'not_empty_in',
+    'AccumBounds', 'is_convex', 'stationary_points', 'minimum', 'maximum',
+
+    'Quaternion',
+
+    'pager_print', 'pretty', 'pretty_print', 'pprint', 'pprint_use_unicode',
```
Comment (asmeurer) [inline sympy/__init__.py]: I would either put these beneath each import, or at least add comments like `# sympy.printing` for each one.

**Comment by asmeurer** [comment]
I guess if the Travis build passes we should remove it from the allowed failures. 

We may also want some documentation on how to skip legitimate false positives, and what kinds of things should be considered legitimate.

**Comment by oscarbenjamin** [comment]
> if the Travis build passes

The travis build won't pass until Py2 is dropped because of the use of the typing module.

Dropping Py2 is in #18243

**Comment by ethankward** [comment]
> We may also want some documentation on how to skip legitimate false positives, and what kinds of things should be considered legitimate.

The main uses of `noqa` in the codebase right now involve injection of variables and `@dispatch` methods.

It's probably also worth looking at if there are any other flake8 errors or warnings we might want to enable.

**Comment by oscarbenjamin** [comment]
The injection functions should be deprecated. For dispatch I would recommend using differently named functions. (I notice that mypy also complains about both of these)

Do you have any suggestions for other useful warnings?

**Comment by ethankward** [comment]
I think we can get rid of `ignore = F403`.

**Comment by asmeurer** [comment]
I don't think we should deprecate the injection functions (I'm assuming you mean `var`), but they definitely shouldn't be used anywhere in library code.

**Comment by oscarbenjamin** [comment]
There are others besides `var` e.g. `vring`:
https://github.com/sympy/sympy/blob/5b87eacfd2b979faaf42b0a450577c4c737d027b/sympy/polys/tests/test_injections.py#L21-L44

**Comment by asmeurer** [comment]
I guess the polys ones aren't that important, although those tests can be rewritten like test_var.py.

**Comment by asmeurer** [comment]
We could replace some of the code quality tests with flake8 warnings. I don't know if it supports everything we check for. 

We should also update our development guides to say that you should run `flake8` before committing. Maybe even `test_code_quality.py` could do it, so that bin/test always runs the checks. 

Most of the [other warnings](https://pep8.readthedocs.io/en/latest/intro.html#error-codes) are code style things which I don't think we should check for.

**Comment by asmeurer** [comment]
> Maybe even test_code_quality.py could do it, so that bin/test always runs the checks.

Although thinking about it this could be a big enough change that it should probably go in a separate PR.

**Comment by asmeurer** [comment]
With explicit imports in the main `__init__.py` we now have the possibility to add an import to a lower `__init__.py` but not the top-level one. This would usually be a typo, so we should add some test that this doesn't happen.

**Comment by oscarbenjamin** [comment]
Do we want all imports at package level to propagate to the main sympy namespace?

A different way of looking at it is that adding explicit imports introduces the possibility of deciding which things should be at top-level and which should only be at package level.

Some packages already aren't imported to top-level (e.g. stats, combinatorics) apparently for performance reasons. I would argue for API that it would be better to stop adding things to the top level (as this PR shows there is already far too much there).

**Comment by asmeurer** [comment]
That's how we've done it so far. If a submodule is already being imported, adding names to `__init__.py` doesn't affect performance. And for any module that's in `sympy/__init__.py` doing a direct import is going to import all of SymPy anyway due to how circular SymPy's imports are.

**Comment by oscarbenjamin** [comment]
Having different things at package level makes it possible to reuse the same name in different contexts e.g. `from sympy.core import E` vs `from sympy.stats import E`.

Performance is not the main issue but I do think it would be better if importing sympy didn't import *everything*. Part of the reason I just moved the test code out of utilities is that there's no way to import anything from utilities without importing *all* of utilities which then imports all of everything else.

**Comment by oscarbenjamin** [comment]
I think a lot of the circularity in SymPy's imports comes from the fact that everything is imported everywhere. Each package `__init__` imports all modules and the main `__init__` imports all of those. It only takes one import that is out of order with that global import chain to create a circular import problem.

**Comment by asmeurer** [comment]
I added some "fixes" and a release notes entry to the OP.

**Comment by oscarbenjamin** [comment]
Another two errors flagged by flake8 snuck in since the tests last ran (yesterday). I've just pushed a fix for those.

Thanks @asmeurer for adding the fixes and improving the release note. From my end this is good to go if the tests pass.

There are some future TODOs:
1. Decide how to handle imports at package vs top-level (or add tests to enforce that everything goes to top-level).
2. Update the dev guide with instructions for running flake8 (or just generally update it).

**Comment by asmeurer** [comment]
OK looks good to me too.

**Comment by oscarbenjamin** [comment]
Thanks. Okay, merging. New PRs will need to match the flake8 rules currently specified.

**Comment by ketch** [comment]
I'm getting this deprecation message even though I don't have `from sympy import *` anywhere in my code.  Is there a reason for that?

**Comment by oscarbenjamin** [comment]
I assume you mean the warning added in #19316. If so then the full warning should tell you what it is that you are accessing e.g. this warning tells me to use `sympy.core.add` instead of `sympy.add`:
```
$ python -c "import sympy; sympy.add.Add"
/Users/enojb/current/sympy/sympy/sympy/__init__.py:672: SymPyDeprecationWarning: 

importing sympy.core.add with 'from sympy import *' has been
deprecated since SymPy 1.6. Use import sympy.core.add instead. See
https://github.com/sympy/sympy/issues/18245 for more info.

  self.Warn(
```
I guess the wording could be improved because it also happens if you do `from sympy import add` rather than `from sympy.core import add`.

**Comment by ketch** [comment]
Thanks for replying.

> the wording could be improved

Yes, based on the current wording I would not have thought that `from sympy import add` is a problem at all.

However, with the additional information you've given, I still can't see why I'm getting this warning:

```
importing sympy.core.basic with 'from sympy import *' has been
deprecated since SymPy 1.6. Use import sympy.core.basic instead.
```

I don't have `from sympy import *` or `from sympy import core` or any other import statement involving
 `basic`.  The only import involving `core` is `import sympy.core.numbers`, which seems to be in the recommended format.

Is there some other phrase I should grep for?

You can see the package in question here: https://github.com/ketch/nodepy

The warning is generated several times when I run the tests.

**Comment by oscarbenjamin** [comment]
The problem is with using `basic` so I'd grep for that.

If you run with `python -Werror` or `pytest -Werror` then the warning will become an exception and you'll get a traceback.

**Comment by ketch** [comment]
Yes, as I mentioned I did grep for basic, and there are no import lines that contain the word basic.

I also ran with `-Werror` and I still get the warnings and no exception.

**Comment by oscarbenjamin** [comment]
Strange that `-Werror` doesn't work. Maybe something else is messing with the warnings filter.

Find the code in `sympy/__init__.py` that emits the warning and replace it with `raise Exception` or something.

**Comment by oscarbenjamin** [comment]
I took a look at nodepy and I see that it is because of `sympy.matrices`. In sympy 1.5.1 we have:
```
>>> import sympy
>>> sympy.matrices
<module 'sympy.matrices.matrices' from '/Users/enojb/current/sympy/sympy/sympy/matrices/matrices.py'>
```
What has happened is that `sympy/__init__.py` did `from .matrices import *` which imports everything from the `sympy.matrices` package. The package also has a submodule `sympy.matrices.matrices` which gets imported by this and is accessible as `sympy.matrices` (overwriting the actual `sympy.matrices` package). In future writing `sympy.matrices` should give you the package rather than the submodule. In sympy 1.6.1 it still gives you the submodule but wrapped in a depreaction warning object:
```
>>> import sympy
>>> sympy.matrices
<sympy.DeprecatedImportModule object at 0x103660580>
>>> sympy.matrices.Matrix
/Users/enojb/current/sympy/sympy/sympy/__init__.py:672: SymPyDeprecationWarning: 

importing sympy.matrices.matrices with 'from sympy import *' has been
deprecated since SymPy 1.6. Use import sympy.matrices.matrices
instead. See https://github.com/sympy/sympy/issues/18245 for more
info.

  self.Warn(
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/enojb/current/sympy/sympy/sympy/__init__.py", line 677, in __getattr__
    return getattr(self.mod, name)
AttributeError: module 'sympy.matrices.matrices' has no attribute 'Matrix'
```
The expected way to access `Matrix` is simply as `sympy.Matrix` because it is exported as a top-level name from `sympy/__init__.py`. Note that `from sympy.matrices import Matrix` will work without the warning. Personally I think that it should be expected to import things from package level e.g. `from sympy.matrices import Matrix` but that's not the currently documented expectation in sympy which is that everything goes to top-level (`from sympy import Matrix` or `sympy.Matrix`).

The warning is there because at some point `sympy.matrices` will refer to the package rather than its submodule. I can't imagine that that would affect the availability of `Matrix` as an attribute but other attributes of the matrices submodule might not be available as public attributes of the matrices package when that happens.

**Comment by ketch** [comment]
@oscarbenjamin Thank you for looking at my code and for the very detailed explanation.  

If I understand you, then my code's sympy imports and in perfect agreement with what the sympy developers want (I always use `sympy.Matrix`), and I shouldn't make any changes.

I hope that this deprecation warning and its trigger will be changed so that it doesn't waste time (like what happened here) for a lot of other people.  It's also a bit annoying to have these "wrong" warnings appear whenever we run the tests.

As a final note, I'm very grateful to the people that donate their time to develop sympy.  But having managed a package that relies on sympy for the last 10 years, I have to say it has been by far the most problematic dependency.  The structure and intended usage are a fast-moving and rather unclear target, as this episode demonstrates.

**Comment by oscarbenjamin** [comment]
@ketch I don't think that you have understood. The expected way to access Matrix is like this:
```
from sympy import Matrix
M = Matrix(...)
```
or like this:
```
import sympy
M = sympy.Matrix(...)
```
What is happening in `nodepy` is this:
```
import sympy
M = sympy.matrices.Matrix(...)
```
That is why you see the warning. If you change it to one of the others then you won't see the warning.

**Comment by asmeurer** [comment]
This looks like an unintended consequence of the deprecation warning to me. `sympy.matrices` used to incorrectly give `sympy.matrices.matrices`. This behavior has been maintained with a deprecation warning. Once the warning is removed, this will give the expected submodule `sympy.matrices`, which will contain the Matrix object. The previous behavior was wrong because `sympy.matrices.matrices` is too specific for most matrix objects. For example, `sympy.matrices.BlockMatrix` currently doesn't work but it should. 

With that being said, it is generally preferred to not rely on specific submodules to import objects, unless they are not objects that are included in `from sympy import *`. We reserve the right to move public objects around in the library without considering it to be a compatibility break.

**Comment by oscarbenjamin** [comment]
This is an unintended consequence of the deprecation warnings. I would prefer to remove the warnings (#19554).

We should have a better separation between what is public and documented and what is internal
https://github.com/sympy/sympy/pull/19554#issuecomment-645914859

**Comment by ketch** [comment]
I finally tracked down the code actually causing the warning.  Although there are some lines with `sympy.matrices.Matrix` in a Jupyter notebook in the examples directory, there are none in the library code, so that isn't what caused the warnings when running the tests.

The offending code was

`import sympy`

followed by a reference to

`sympy.basic.Basic`

What threw me off and made it hard to find this was that the warning seemed to claim there was something wrong with an import statement, which isn't the case.  So perhaps it could be reworded to be more clear.  In any case, thanks for all the help!

**Comment by asmeurer** [comment]
I think we didn't realize that people we using SymPy like this when we wrote the warning. Most likely the next version of SymPy will just remove the warnings and compatibility completely.

**Comment by asmeurer** [comment]
(by the way, I just realized people are commenting on this old PR because the issue number in the code is wrong)

**Comment by oscarbenjamin** [comment]
I think that for the next release we can at least remove all the cases that conflict with an actual package i.e. logic, integrals, matrices, sets and solvers. @ketch didn't need to see any warning here if `sympy.matrices` had simply referred to the correct package rather than still referring to the old `sympy.matrices.matrices` module but with a warning. The other warnings are harmless to leave there for now.

What this shows though:

1. It's definitely good to get all the imports cleaned up as this PR (and all the others) did.

2. Users are accessing things from the internals of sympy so moving things around the codebase will break downstream code.

3. It would be better to have a clearly separate package structure for the public API than for the internal code.

4. Users expect that there should be some namespacing within the sympy API and really there should. We should stop putting everything to top-level.

**Comment by asmeurer** [comment]
> (by the way, I just realized people are commenting on this old PR because the issue number in the code is wrong)

Or perhaps this was intentional, since this was the PR that added `__all__` to `__init__.py`? It would be better for deprecation issues to be dedicated issues, rather than the PR that removed the issue, that way they can describe the motivation for the deprecation and how to fix user code without being mixed up with PR technicalities.

**Comment by asmeurer** [comment]
@ketch I can see how you would end up with `sympy.matrices.Matrix`. The fact that the submodule is the same name as the module makes `sympy.matrices.matrices` impossible (at least until we remove this deprecation). But I'm curious how you ended up with `sympy.basic.Basic`? Was it by mimicking the sympy.matrices.Matrix`? Did you find it by accident or is that copied from somewhere? And how did you end up trying it first rather than just `sympy.Basic`?

**Comment by oscarbenjamin** [comment]
> It would be better for deprecation issues to be dedicated issues

There's an awkward boot strap problem because you don't know the issue number to put into the code until after you create the pull request or otherwise you create the issue before the details of the change are agreed.

**Comment by asmeurer** [comment]
Yeah, you have to make a separate issue before the PR is merged. If there's uncertainty about whether the deprecation will be done in the issue, you can hold off on creating it until there is consensus, then create it before merging the PR.

**Comment by ketch** [comment]
> @ketch I can see how you would end up with `sympy.matrices.Matrix`. The fact that the submodule is the same name as the module makes `sympy.matrices.matrices` impossible (at least until we remove this deprecation). But I'm curious how you ended up with `sympy.basic.Basic`? Was it by mimicking the `sympy.matrices.Matrix`? Did you find it by accident or is that copied from somewhere? And how did you end up trying it first rather than just `sympy.Basic`?

Like I said, throughout the library I actually use `sympy.Matrix` and not `sympy.matrices.Matrix`.

The one line with `sympy.basic.Basic` was committed in 2012, when a sympy change broke the previous version which was `sympy.core.basic.Basic`.  If I recall correctly, I was just trying to find where the `Basic` object had been moved to and that was the first thing I found.  I should have said before -- I am happy that nowadays there are deprecation warnings prior to something being moved or removed!  I think that in 2012 it just suddenly broke with no warning.  But that might have been before sympy v1.0, so it was understandable.

**Comment by asmeurer** [comment]
I see. That makes sense. Actually the change that this is deprecating is exactly the sort of thing that originally broke `sympy.core.basic.Basic`. It would have stopped working when `sympy.core` actually referred to `sympy.core.core`, making the top-level `sympy.core` module almost impossible to access. Once this deprecation is removed, all modules will be accessible as expected, even if their names are duplicated.

**Comment by oscarbenjamin** [comment]
I think that the changes introduced in this PR (and many related PRs) will help prevent breakage in future because imports through all packages are explicit. We don't use `import *` anywhere in the codebase so we have explicit control of when accessibility of imported names is changed provided you always access them at package level or from top-level.

I have a PR that will fix this warning for matrices and a few other packages (#19775) so that `sympy.matrices` will be the correct package rather than the incorrect submodule with a warning. There will have been a warning for one release (1.6) and then we can fix it so the imports for those cases work as they should (in 1.7). We can probably leave the other 60 warnings there for now as removing those would lead to ImportError. The cases changed in #19775 are bugs that remain unfixed because the warnings were left in place.

**Pull request merged**